plugin/src/Machinery.cc

```diff
@@ -3758,6 +3758,49 @@ Do_var::load(Config &cfg, CfgStaticData const *, YAML::Node, swoc::TextView cons
 
   return Handle(new self_type(cfg.localize(arg), std::move(expr)));
 }
+/* ------------------------------------------------------------------------------------ */
+/// Internal transaction error control
+class Do_txn_error : public Directive
+{
+  using self_type  = Do_txn_error;
+  using super_type = Directive;
+
+public:
+  static inline const std::string KEY{"txn-error"};
+  /// Valid hooks for directive.
+  static inline const HookMask HOOKS{MaskFor({Hook::TXN_START, Hook::CREQ, Hook::PREQ, Hook::URSP, Hook::PRSP,
+                                              Hook::PRE_REMAP, Hook::POST_REMAP, Hook::REMAP})};
+
+  Errata invoke(Context &ctx) override;
+  static Rv<Handle> load(Config &cfg, CfgStaticData const *, YAML::Node drtv_node, swoc::TextView const &name,
+                         swoc::TextView const &arg, YAML::Node key_value);
+
+protected:
+  Expr _expr;
+
+  Do_txn_error(Expr &&msg);
+};
+
+Do_txn_error::Do_txn_error(Expr &&expr) : _expr(std::move(expr)) {}
+
+Errata
+Do_txn_error::invoke(Context &ctx)
+{
+  ctx._global_status = ctx.extract(_expr).as_bool() ? TS_EVENT_HTTP_ERROR : TS_EVENT_HTTP_CONTINUE;
+  return {};
```

Comment:
unrelated question, but why the return is decoupled from the logic? I mean, seems that most of the time we return no error.